pipeline/metadata/flatten_satellite.py

```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
-  row['matches_confidence'] = scan.get('confidence')['matches']
-  row['untagged_controls'] = scan.get('confidence')['untagged_controls']
-  row['untagged_response'] = scan.get('confidence')['untagged_response']
-
-  row['excluded'] = scan.get('excluded', False)
-  row['exclude_reason'] = ' '.join(scan.get('exclude_reason', []))
-
-  errors = [
-      response['error']
-      for response in responses
-      if response['error'] and response['error'] != 'null'
-  ]
-  row['error'] = ' | '.join(errors) if errors else None
-  yielded = False
-  has_test_domain = False
-  for response in responses:
-    if response['url'] == row['domain']:
-      has_test_domain = True
-    # Check responses for test domain with valid answers
-    if response['url'] == row['domain'] and (response['rcode'] == 0 and
-                                             response['has_type_a']):
-      row['has_type_a'] = True
-      # Separate into one answer IP per row for tagging
-      row['received'] = []
-      for ip in response['response']:
-        received = {
-            'ip': ip,
-            'http': response['response'][ip].get('http'),
-            'cert': response['response'][ip].get('cert'),
-            'asname': response['response'][ip].get('asname'),
-            'asnum': response['response'][ip].get('asnum'),
-            'matches_control': ''
-        }
-        matched = response['response'][ip].get('matched', [])
-        if matched:
-          received['matches_control'] = ' '.join(
-              [tag for tag in matched if tag in SATELLITE_TAGS])
-        row['received'].append(received)
-      yielded = True
-      yield row.copy()
-    # Do not output rows if there are no trials for the test domain.
-  if not yielded and has_test_domain:
-    # All trials for the test domain in `scan` are errors
-    yield row.copy()
+  row['controls_failed'] = not scan['passed_control']
+  received_ips = scan.get('response').copy()
+
+  if not received_ips:
+    row['received'] = []
+    yield row
+    return
+
+  # separate into one answer ip per row for tagging
+  rcodes = received_ips.pop('rcode', [])
+  errors = received_ips.pop('error', [])
+
+  successful_test_rcode, failed_test_rcodes = split_rcodes(rcodes)
+
+  if any(rcode == "0" for rcode in failed_test_rcodes):
+    raise Exception(
+        f"Satellite v2.1 measurement has multiple 0 rcodes: {filepath} - {scan}"
+    )
+
+  if not successful_test_rcode and received_ips:
+    # TODO figure out how to handle this bug case from 2021-05-16
+    # for now drop the measurements
+    logging.warning(
+        f"Satellite v2.1 measurement has ips but no 0 rcode: {filepath} - {scan}"
+    )
+    return
+
+  if successful_test_rcode:
+    all_received = []
+    for (ip, tags) in received_ips.items():
+      received = {'ip': ip}
+      if tags:
+        received['matches_control'] = _append_tags(tags)
+      all_received.append(received)
+
+    success_row = row.copy()
```

Comment:
Rename `success_row` to `observation`. It should represent a single DNS request + observation.

Reply:
renamed to `success_observation`. That is what it represents.

Reply:
renamed `row` to `base_observation` and `received_ips` to `input_response_data`